From: Stefan Assmann <sassmann@redhat.com>
Date: Fri, 22 Oct 2010 17:53:13 -0400
Subject: [net] igb: fix TX hang when loading igb with max_vfs > 7
Message-id: <4CC1CF89.50402@redhat.com>
Patchwork-id: 28901
O-Subject: [RHEL 5.6 PATCH] Fix Tx hangs seen when loading igb with max_vfs > 7
Bugzilla: 645284
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Andy Gospodarek <gospo@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

Bugzilla:
https://bugzilla.redhat.com/show_bug.cgi?id=645284

Description:
Check the value of max_vfs at the time of assignment of vfs_allocated_count.

The previous check in igb_probe_vfs was too late as by that time the rx/tx
rings were initialized with the wrong offset.

Upstream Status:
http://git.kernel.org/linus/c0f2276f3601a96b29d4347c5938e9494e8e4e5d

Brew Build:
http://brewweb.devel.redhat.com/brew/taskinfo?taskID=2846207

Test Status:
Verified that the patch allows igb to work with max_vfs > 7.

  Stefan

diff --git a/drivers/net/igb/igb_main.c b/drivers/net/igb/igb_main.c
index 57c2db9..ec87c95 100644
--- a/drivers/net/igb/igb_main.c
+++ b/drivers/net/igb/igb_main.c
@@ -2079,9 +2079,6 @@ static void __devinit igb_probe_vfs(struct igb_adapter * adapter)
 #ifdef CONFIG_PCI_IOV
 	struct pci_dev *pdev = adapter->pdev;
 
-	if (adapter->vfs_allocated_count > 7)
-		adapter->vfs_allocated_count = 7;
-
 	if (adapter->vfs_allocated_count) {
 		adapter->vf_data = kcalloc(adapter->vfs_allocated_count,
 		                           sizeof(struct vf_data_storage),
@@ -2139,7 +2136,7 @@ static int __devinit igb_sw_init(struct igb_adapter *adapter)
 
 #ifdef CONFIG_PCI_IOV
 	if (hw->mac.type == e1000_82576)
-		adapter->vfs_allocated_count = max_vfs;
+		adapter->vfs_allocated_count = (max_vfs > 7) ? 7 : max_vfs;
 
 #endif /* CONFIG_PCI_IOV */
 	/* This call may decrease the number of queues */
